gpt_engineer/applications/cli/cli_agent.py

```diff
@@ -0,0 +1,127 @@
+from gpt_engineer.core.default.steps import (
+    gen_code,
+    gen_entrypoint,
+    execute_entrypoint,
+    improve,
+)
+from gpt_engineer.core.repository import Repository
+from gpt_engineer.core.default.on_disk_repository import OnDiskRepository
+from gpt_engineer.core.execution_env import ExecutionEnv
+from gpt_engineer.core.default.on_disk_execution_env import OnDiskExecutionEnv
+from gpt_engineer.core.default.paths import memory_path, ENTRYPOINT_FILE, PREPROMPTS_PATH
+from gpt_engineer.core.agent import Agent
+from gpt_engineer.core.preprompts_holder import PrepromptsHolder
+from typing import TypeVar, Callable, Union
+from pathlib import Path
+
+CodeGenType = TypeVar("CodeGenType", bound=Callable[[AI, str, Repository], Code])
+CodeProcessor = TypeVar("CodeProcessor", bound=Callable[[AI, ExecutionEnv, Code], Code])
+ImproveType = TypeVar("ImproveType", bound=Callable[[AI, str, Code, Repository], Code])
+
+
+class CliAgent(Agent):
+    """
+    The `Agent` class is responsible for managing the lifecycle of code generation and improvement.
+
+    Attributes:
+        path (str): The file path where the `Agent` will operate, used for version management and
+                    file operations.
+        version_manager (BaseVersionManager): An object that adheres to the VersionManagerInterface,
+                        responsible for version control of the generated code. Defaults to `VersionManager`
+                        if not provided. PROBABLY GIT SHOULD BE USED IN THE DEFAULT
+        step_bundle (StepBundleInterface): Workflows of code generation steps that define the behavior of gen_code and
+        improve.
+        ai (AI): Manages calls to the LLM.
+
+    Methods:
+        __init__(self, path: str, version_manager: VersionManagerInterface = None,
+                 step_bundle: StepBundleInterface = None, ai: AI = None):
+            Initializes a new instance of the Agent class with the provided path, version manager,
+            step bundle, and AI. It falls back to default instances if specific components are not provided.
+
+        init(self, prompt: str) -> Code:
+            Generates a new piece of code using the AI and step bundle based on the provided prompt.
+            It also snapshots the generated code using the version manager.
+
+            Parameters:
+                prompt (str): A string prompt that guides the code generation process.
+
+            Returns:
+                Code: An instance of the `Code` class containing the generated code.
+
+        improve(self, prompt: str) -> Code:
+            Improves an existing piece of code using the AI and step bundle based on the provided prompt.
+            It also snapshots the improved code using the version manager.
+
+            Parameters:
+                prompt (str): A string prompt that guides the code improvement process.
+
+            Returns:
+                Code: An instance of the `Code` class containing the improved code.
```

Comment:
I'm curious: why aren't these generated with things like [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html)?

The project has a lot of docstrings which seem a pain to manually maintain and keep up-to-date, which autodoc would simplify massively (at the cost of not getting it exactly how you want it).

Reminds me of [this video](https://www.youtube.com/watch?v=k5DJxz6_xMA)

Reply:
Would be amazing if you could have a shot at this once we merge.